EchoPro/computation/statistics.py

```diff
@@ -0,0 +1,128 @@
+    strata: pd.DataFrame
+        DataFrame comprising summary features of latitude (INPFC) delimited strata
+    sample_fraction: np.float64
+        Value representing the proportion of ftransects that are resampled from the
+        overall dataset within each strata
+    replicates: int
+        The number of iterations/realizations used for bootstrapping
+
+    Notes
+    -----
+    This function calculates the stratified summary statistics for biomass within 
+    `EchoPro.survey.stratified_summary()`.
+    """ 
+    
+    ### Convert specific DataFrame columns to arrays for speed
+    distance = transects[ 'transect_distance' ].values
+    value = transects[ 'B_adult' ].values
+    num_transects = strata[ 'num_transects' ].values
+    total_transect_area = strata.set_index( 'stratum_inpfc' )[ 'total_transect_area' ]
+
+    ### Calculate the number of transects within each stratum based on the 
+    ### sampling faction defined from the configuration file
+    # Number of transects
+    num_transects_to_sample = np.round( sample_fraction * num_transects ).astype( int )
+
+    # Offset term used for later variance calculation
+    sample_offset = np.where( num_transects_to_sample == 1 , 0 , 1 )
+
+    # Calculate effective sample size/degrees of freedom for variance calculation
+    sample_dof =  num_transects_to_sample * ( num_transects_to_sample - sample_offset )
+
+    ### Pre-allocate and pre-compute the cumulative sum of numbered transects per strata
+    # Transect indices
+    cum_num_transects = np.concatenate( ( [ 0 ] , np.cumsum( num_transects ) ) )
+
+    # Stratified statistics
+    mean_arr = np.empty( replicates )
+    variance_arr = np.empty( replicates )
+    
+    ### Iterate across all replicate iterations/realizations
+    for i in range( replicates ):
+
+        # Pre-allocate the stratum-specific means and variances
+        rho_j = np.empty_like( total_transect_area ) # mean
+        var_j = np.empty_like( total_transect_area ) # variance
+
+        # Iterate across all strata
+        for j in strata.stratum_inpfc - 1:
+
+            ### Resample (without replacement) based on binned indices
+            # Define start and end transects within each stratum
+            start , end = cum_num_transects[ j ] , cum_num_transects[ j + 1 ]
+
+            # Resample without replacement
+            sel_inds = np.random.choice( np.arange( start , end ) , 
+                                         num_transects_to_sample[ j ] , 
+                                         replace=False )
+            
+            ### Define stratified weights
+            stratified_weights = distance[ sel_inds ] / np.mean( distance[ sel_inds ] )
+
+            ### Weighted value (e.g. biomass)
+            value_distance_density = value[ sel_inds ] / distance[ sel_inds ]
+
+            ### Compute mean and variance
+            rho_j[ j ] = np.nansum( value[ sel_inds ] * stratified_weights ) / np.nansum( stratified_weights )
```

Comment:
In what case would `stratified_weights` be NaN? It would have to come from `distance[ sel_inds ]`, so I guess if `distance` came from a diff operation, there could be NaN, but I have not tried to trace the code further up.

Reply:
When developing I think this was the case, but adjustments to `calculate_transect_distance` may deprecate the need for tip-toeing around NaN values here. I know that `numpy.nansum` adds computational costs -- so this may be worth fully evaluating later on. I'll add this as an issue.